### PR TITLE
Feature/tftrainer

### PR DESCRIPTION
**Pull request by crystina-z**
### changes:
1. change the meaning of `itersize`  in tensorflow trainer from "number of steps" to "number of instance" (as mentioned in #117 )
2. add fast-forward training to tensorflow trainer (review required)
3. rename `warmupsteps` -> `warmupiters`, `decaystep` -> `decayiters`
4. change the criteria of determinating "bert" variable, to also include "electra"

#### before the changes (master branch, `bf50423`)
| Command (`python -m capreolus.run rerank.train with ...`) | NDCG@20 |
|-----------------------------------------------------------------------------------------------------------------------------------|-------|
| `... reranker.name=KNRM reranker.trainer.name=pytorch` | 0.40  |
| `... reranker.name=TFKNRM reranker.trainer.name=tensorflow ` | 0.398 | 
| `... reranker.name=TFKNRM reranker.trainer.name=tensorflow reranker.trainer.fastforward=True reranker.itersize=16` |  0.312   | 

#### after the changes (`b883e11`)
| Command (`python -m capreolus.run rerank.train with reranker.trainer.fastforward=True...`) | NDCG@20 |
|-----------------------------------------------------------------------------------------------------------------------------------|-------|
| `... reranker.name=KNRM reranker.trainer.name=pytorch` | 0.40  |
| `... reranker.name=TFKNRM reranker.trainer.name=tensorflow` | 0.353 | 
| `... reranker.name=TFKNRM reranker.trainer.name=tensorflow reranker.trainer.itersize=16384` |  0.395 |

### CODE REVIEW

File: capreolus/trainer/__init__.py [hunk 1]
```diff
@@ -16,11 +16,54 @@ class Trainer(ModuleBase):
...
+    def n_batch_per_iter(self):
+        return (self.config["itersize"] // self.config["batch"]) or 1
+
+
     def get_paths_for_early_stopping(self, train_output_path, dev_output_path):
-        os.makedirs(dev_output_path, exist_ok=True)
         dev_best_weight_fn = train_output_path / "dev.best"
         weights_output_path = train_output_path / "weights"
         info_output_path = train_output_path / "info"
+        os.makedirs(dev_output_path, exist_ok=True)
```
Comment (andrewyates) [inline capreolus/trainer/__init__.py:66]: indentation?

**Comment by andrewyates** [comment]
Could you also add the mixed precision changes to this PR? (they should be easy to integrate, right?)

I'm a bit confused about the terminology. `itersize` is the number of samples in an epoch, right? And `warmupiters` and `decayiters` are specified in batches, not iterations? I think the previous `warmupstep` naming would be more clear in that case.

**Comment by crystina-z** [comment]
> @crystina-z could you also add the mixed precision changes to this PR? (they should be easy to integrate, right?)

yeah I've moved the mixed precision code used for bert to this branch (haven't committed and pushed) but then find with that the loss of `TFKNRM` becomes nan... let's switch to slack to discuss this?

> I'm a bit confused about the terminology. `itersize` is the number of samples in an epoch, right? And `warmupiters` and `decayiters` are specified in batches, not iterations? I think the previous `warmupstep` naming would be more clear in that case.

actually `warmupiters` and `decayiters` are also specified in an epoch (I'm assuming epoch and iteration are used interchangeably here?) but the warmup and decay are done in batch (or step) granularity ([here](https://github.com/capreolus-ir/capreolus/blob/79e927e38f7873d88e0d326451901aabb839cbdb/capreolus/trainer/__init__.py#L98-L99)). 
Example: if we set `niters=10 itersize=512 batchsize=16 warmupiters=2 decayiters=8`, that will mean 
1. the whole training involves 10 epochs with 512 instances (or 512/16=32 batches) per epoch
2. on the first two epoches, the warmup is applied every step (32 times per epoch)
3. similarity, on the last 8 epoches, the decay is applied every step (32 times per epoch)

**Comment by andrewyates** [comment]
Merging into another branch first, so tests will run and I can push a ConfigOption for AMP.
